File: rust/contacts/src/legacy_kit_models.rs
use ente_core::{auth::KeyAttributes, crypto::SecretVec};
use serde::{Deserialize, Deserializer, Serialize, Serializer, de};

pub const LEGACY_KIT_PAYLOAD_VERSION: u8 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LegacyKitVariant {
    TwoOfThree,
}

impl LegacyKitVariant {
    pub const TWO_OF_THREE_CODE: u8 = 1;

    pub fn code(self) -> u8 {
        match self {
            Self::TwoOfThree => Self::TWO_OF_THREE_CODE,
        }
    }

    pub fn threshold(self) -> usize {
        match self {
            Self::TwoOfThree => 2,
        }
    }

    pub fn part_count(self) -> usize {
        match self {
            Self::TwoOfThree => 3,
        }
    }

    pub fn from_code(code: u8) -> Option<Self> {
        match code {
            Self::TWO_OF_THREE_CODE => Some(Self::TwoOfThree),
            _ => None,
        }
    }
}

impl Serialize for LegacyKitVariant {
    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_u8(self.code())
    }
}

impl<'de> Deserialize<'de> for LegacyKitVariant {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let code = u8::deserialize(deserializer)?;
        Self::from_code(code).ok_or_else(|| de::Error::custom("invalid legacy kit variant"))
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum LegacyKitRecoveryStatus {
    Waiting,
    Ready,
    Blocked,
    Cancelled,
    Recovered,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitRecoverySession {
    pub id: String,
    #[serde(rename = "kitID")]
    pub kit_id: String,
    pub status: LegacyKitRecoveryStatus,
    pub wait_till: i64,
    pub created_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitRecoveryInitiator {
    #[serde(default)]
    pub used_part_indexes: Vec<u8>,
    pub ip: String,
    pub user_agent: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitOwnerRecoverySession {
    pub session: Option<LegacyKitRecoverySession>,
    pub initiators: Vec<LegacyKitRecoveryInitiator>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitPart {
    pub index: u8,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitMetadata {
    pub parts: Vec<LegacyKitPart>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKit {
    pub id: String,
    pub variant: LegacyKitVariant,
    pub notice_period_in_hours: i32,
    pub metadata: LegacyKitMetadata,
    pub created_at: i64,
    pub updated_at: i64,
    pub active_recovery_session: Option<LegacyKitRecoverySession>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct LegacyKitShare {
    #[serde(rename = "pv")]
    pub payload_version: u8,
    #[serde(rename = "kv")]
    pub variant: LegacyKitVariant,
    #[serde(rename = "k")]
    pub kit_id: String,
    #[serde(rename = "i")]
    pub share_index: u8,
    #[serde(rename = "s")]
    pub share: String,
    #[serde(rename = "c")]
    pub checksum: String,
    #[serde(rename = "n")]
    pub part_name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct LegacyKitCreateResult {
    pub kit: LegacyKit,
    pub shares: Vec<LegacyKitShare>,
}

#[derive(Debug)]
pub struct LegacyKitRecoveryBundle {
    pub recovery_key: SecretVec,
    pub user_key_attributes: KeyAttributes,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn legacy_kit_share_serializes_with_compact_qr_keys() {
        let share = LegacyKitShare {
            payload_version: LEGACY_KIT_PAYLOAD_VERSION,
            variant: LegacyKitVariant::TwoOfThree,
            kit_id: "kit-id".into(),
            share_index: 1,
            share: "share".into(),
            checksum: "checksum".into(),
            part_name: "North".into(),
        };

        let json = serde_json::to_string(&share).unwrap();
        assert_eq!(
            json,
            r#"{"pv":1,"kv":1,"k":"kit-id","i":1,"s":"share","c":"checksum","n":"North"}"#
        );
        let decoded: LegacyKitShare = serde_json::from_str(&json).unwrap();
        assert_eq!(decoded, share);
    }

    #[test]
    fn legacy_kit_share_rejects_unknown_variant() {
        let result = serde_json::from_str::<LegacyKitShare>(
            r#"{"pv":1,"kv":9,"k":"kit-id","i":1,"s":"share","c":"checksum","n":"North"}"#,
        );
        assert!(result.is_err());
    }
}

File: mobile/packages/legacy/lib/models/legacy_kit_models.dart
import "dart:convert";

import "package:ente_rust/ente_rust.dart" as rust;

const String defaultLegacyRecoveryUrl = "https://legacy.ente.com";

enum LegacyKitRecoveryStatus {
  waiting,
  ready,
  blocked,
  cancelled,
  recovered,
}

class LegacyKitRecoverySession {
  final String id;
  final String kitId;
  final LegacyKitRecoveryStatus status;
  // Remaining microseconds until recovery becomes ready, matching the existing
  // legacy contact recovery API contract.
  final int waitTill;
  final int createdAt;

  const LegacyKitRecoverySession({
    required this.id,
    required this.kitId,
    required this.status,
    required this.waitTill,
    required this.createdAt,
  });

  factory LegacyKitRecoverySession.fromRust(
    rust.LegacyKitRecoverySession session,
  ) {
    return LegacyKitRecoverySession(
      id: session.id,
      kitId: session.kitId,
      status: _statusFromRust(session.status),
      waitTill: session.waitTill,
      createdAt: session.createdAt,
    );
  }
}

class LegacyKitRecoveryInitiatorHint {
  final List<int> usedPartIndexes;
  final String ip;
  final String userAgent;

  const LegacyKitRecoveryInitiatorHint({
    required this.usedPartIndexes,
    required this.ip,
    required this.userAgent,
  });

  factory LegacyKitRecoveryInitiatorHint.fromRust(
    rust.LegacyKitRecoveryInitiatorHint hint,
  ) {
    return LegacyKitRecoveryInitiatorHint(
      usedPartIndexes: hint.usedPartIndexes,
      ip: hint.ip,
      userAgent: hint.userAgent,
    );
  }
}

class LegacyKitOwnerRecoverySessionDetails {
  final LegacyKitRecoverySession? session;
  final List<LegacyKitRecoveryInitiatorHint> initiators;

  const LegacyKitOwnerRecoverySessionDetails({
    required this.session,
    required this.initiators,
  });

  factory LegacyKitOwnerRecoverySessionDetails.fromRust(
    rust.LegacyKitOwnerRecoverySessionDetails details,
  ) {
    return LegacyKitOwnerRecoverySessionDetails(
      session: details.session == null
          ? null
          : LegacyKitRecoverySession.fromRust(details.session!),
      initiators: details.initiators
          .map(LegacyKitRecoveryInitiatorHint.fromRust)
          .toList(growable: false),
    );
  }
}

class LegacyKitPart {
  final int index;
  final String name;

  const LegacyKitPart({
    required this.index,
    required this.name,
  });

  factory LegacyKitPart.fromRust(rust.LegacyKitPart part) {
    return LegacyKitPart(index: part.index, name: part.name);
  }
}

class LegacyKit {
  final String id;
  final int noticePeriodInHours;
  final String legacyUrl;
  final List<LegacyKitPart> parts;
  final int createdAt;
  final int updatedAt;
  final LegacyKitRecoverySession? activeRecoverySession;

  const LegacyKit({
    required this.id,
    required this.noticePeriodInHours,
    required this.legacyUrl,
    required this.parts,
    required this.createdAt,
    required this.updatedAt,
    required this.activeRecoverySession,
  });

  factory LegacyKit.fromRust(rust.LegacyKit kit) {
    return LegacyKit(
      id: kit.id,
      noticePeriodInHours: kit.noticePeriodInHours,
      legacyUrl: kit.legacyUrl.trim().isEmpty
          ? defaultLegacyRecoveryUrl
          : kit.legacyUrl,
      parts: kit.metadata.parts.map(LegacyKitPart.fromRust).toList(
            growable: false,
          ),
      createdAt: kit.createdAt,
      updatedAt: kit.updatedAt,
      activeRecoverySession: kit.activeRecoverySession == null
          ? null
          : LegacyKitRecoverySession.fromRust(kit.activeRecoverySession!),
    );
  }

  String get displayName => parts.map((part) => part.name).join(" · ");

  bool get hasActiveRecoverySession => activeRecoverySession != null;
}

class LegacyKitShare {
  final int payloadVersion;
  final int variant;
  final String kitId;
  final int shareIndex;
  final String share;
  final String checksum;
  final String partName;

  const LegacyKitShare({
    required this.payloadVersion,
    required this.variant,
    required this.kitId,
    required this.shareIndex,
    required this.share,
    required this.checksum,
    required this.partName,
  });

  factory LegacyKitShare.fromRust(rust.LegacyKitShare share) {
    return LegacyKitShare(
      payloadVersion: share.payloadVersion,
      variant: _variantToCode(share.variant),
      kitId: share.kitId,
      shareIndex: share.shareIndex,
      share: share.share,
      checksum: share.checksum,
      partName: share.partName,
    );
  }

  String toQrPayload() {
    return jsonEncode({
      "pv": payloadVersion,
      "kv": variant,
      "k": _withoutWhitespace(kitId),
      "i": shareIndex,
      "s": _withoutWhitespace(share),
      "c": _withoutWhitespace(checksum),
      "n": partName,
    });
  }

  String toCopyCode() {
    return base64Url.encode(utf8.encode(toQrPayload())).replaceAll("=", "");
  }

  String _withoutWhitespace(String value) {
    return value.replaceAll(RegExp(r"\s+"), "");
  }
}

class LegacyKitCreateResult {
  final LegacyKit kit;
  final List<LegacyKitShare> shares;

  const LegacyKitCreateResult({
    required this.kit,
    required this.shares,
  });

  factory LegacyKitCreateResult.fromRust(rust.LegacyKitCreateResult result) {
    return LegacyKitCreateResult(
      kit: LegacyKit.fromRust(result.kit),
      shares: result.shares.map(LegacyKitShare.fromRust).toList(
            growable: false,
          ),
    );
  }
}

LegacyKitRecoveryStatus _statusFromRust(rust.LegacyKitRecoveryStatus status) {
  return switch (status) {
    rust.LegacyKitRecoveryStatus.waiting => LegacyKitRecoveryStatus.waiting,
    rust.LegacyKitRecoveryStatus.ready => LegacyKitRecoveryStatus.ready,
    rust.LegacyKitRecoveryStatus.blocked => LegacyKitRecoveryStatus.blocked,
    rust.LegacyKitRecoveryStatus.cancelled => LegacyKitRecoveryStatus.cancelled,
    rust.LegacyKitRecoveryStatus.recovered => LegacyKitRecoveryStatus.recovered,
  };
}

int _variantToCode(rust.LegacyKitVariant variant) {
  return switch (variant) {
    rust.LegacyKitVariant.twoOfThree => 1,
  };
}
